lib/tbot/service_ssh_multiplexer.go

```diff
@@ -0,0 +1,751 @@
+	buf := bufio.NewReader(downstream)
+	req, err := buf.ReadString('\x00')
+	if err != nil {
+		return trace.Wrap(err, "reading request")
+	}
+	req = req[:len(req)-1] // Drop the NUL.
+	host, port, err := utils.SplitHostPort(req)
+	if err != nil {
+		return trace.Wrap(err, "malformed request %q", req)
+	}
+
+	log := s.log.With(
+		slog.Group("req",
+			"host", host,
+			"port", port,
+		),
+	)
+	log.InfoContext(ctx, "Received multiplexing request")
+
+	clusterName := s.identity.Get().ClusterName
+	expanded, matched := tshConfig.ProxyTemplates.Apply(
+		net.JoinHostPort(host, port),
+	)
+	if matched {
+		log.DebugContext(
+			ctx,
+			"Proxy templated matched",
+			"populated_template", expanded,
+		)
+		if expanded.Cluster != "" {
+			clusterName = expanded.Cluster
+		}
+
+		if expanded.Host != "" {
+			host = expanded.Host
+		}
+	}
+
+	var target string
+	if expanded == nil || (len(expanded.Search) == 0 && expanded.Query == "") {
+		host = cleanTargetHost(host, proxyHost, clusterName)
+		target = net.JoinHostPort(host, port)
+	} else {
+		node, err := resolveTargetHostWithClient(ctx, authClient, expanded.Search, expanded.Query)
+		if err != nil {
+			return trace.Wrap(err, "resolving target host")
+		}
+
+		log.DebugContext(
+			ctx,
+			"Found matching SSH host",
+			"host_uuid", node.GetName(),
+			"host_name", node.GetHostname(),
+		)
+
+		target = net.JoinHostPort(node.GetName(), "0")
+	}
+
+	upstream, _, err := hostDialer.DialHost(ctx, target, clusterName, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if s.cfg.SessionResumptionEnabled() {
+		log.DebugContext(ctx, "Enabling session resumption")
+		upstream, err = resumption.WrapSSHClientConn(
+			ctx,
+			upstream,
+			func(ctx context.Context, hostID string) (net.Conn, error) {
+				log.DebugContext(ctx, "Resuming connection")
+				// if the connection is being resumed, it means that
+				// we didn't need the agent in the first place
+				var noAgent agent.ExtendedAgent
+				conn, _, err := hostDialer.DialHost(
+					ctx, net.JoinHostPort(hostID, "0"), clusterName, noAgent,
+				)
+				return conn, err
+			})
+		if err != nil {
+			return trace.Wrap(err, "wrapping conn for session resumption")
+		}
+	}
+	defer func() {
+		if err := upstream.Close(); err != nil {
+			s.log.DebugContext(ctx, "Error closing upstream connection", "error", err)
+		}
+	}()
+
+	defer context.AfterFunc(ctx, func() {
+		if err := trace.NewAggregate(
+			upstream.Close(),
+			downstream.Close(),
+		); err != nil {
+			s.log.DebugContext(ctx, "Error closing connections", "error", err)
+		}
+	})()
+
+	log.InfoContext(ctx, "Proxying connection for multiplexing request")
+	startedProxying := time.Now()
+
+	errCh := make(chan error, 2)
+	go func() {
+		defer upstream.Close()
+		defer downstream.Close()
+		// Drain the buffer we used to read in the request in case it read in more
+		// than just the initial request.
+		_, err := io.CopyN(upstream, buf, int64(buf.Buffered()))
+		if err != nil {
```

Comment:
Something I just realized we can do, and it's conceptually simpler than `io.CopyN`:
```suggestion
		b, _ := buf.Peek(buf.Buffered())
		if _, err := upstream.Write(b); err != nil {
```

Reply:
[3258e5d](https://github.com/gravitational/teleport/pull/42592/commits/3258e5dd2e93e2c603a72b951cb3d2829484fb45)